kube-runtime/src/observer.rs

```diff
@@ -0,0 +1,114 @@
+use backoff::ExponentialBackoff;
+use futures::{future::ready, Stream, StreamExt};
+use kube_client::{
+    api::{Api, ListParams},
+    core::Resource,
+};
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+
+use crate::{
+    utils,
+    watcher::{watcher, Error, Event, Result},
+};
+
+
+/// A simple observer around a watcher with error handling and retry backoff
+///
+/// # Error handling
+///
+/// An `Observer` sets a sensible default backoff policy for all watch events and will retry
+/// (with expotential backoff) from transient `watcher` failures until the retry policy is breached.
+///
+/// To configure the retry policy use `Observer::backoff`.
+///
+/// Note that while all errors will be retried, some represent the need for user action to recover:
+///
+/// - 404 `ErrorResponse`(watching invalid / missing api kind/group for `K`)
+/// - 403 `ErrorResponse` (missing list + watch rbac verbs for `K`)
+pub struct Observer<K>
+where
+    K: Clone + Resource + Send + Sync + 'static,
+{
+    // temporary builder params
+    api: Api<K>,
+    listparams: Option<ListParams>,
+    backoff: Option<ExponentialBackoff>,
+}
+
+impl<K> From<Api<K>> for Observer<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+{
+    fn from(api: Api<K>) -> Observer<K> {
+        Observer::new(api)
+    }
+}
+
+impl<K> Observer<K>
+where
+    K: Clone + Resource + DeserializeOwned + Debug + Send + Sync + 'static,
+{
+    /// Create a Observer on a reflector on a type `K`
+    ///
+    /// Takes an [`Api`] object that determines how the `Observer` listens for changes to the `K`.
+    ///
+    /// The [`ListParams`] controls to the possible subset of objects of `K` that you want to cache.
+    /// For the full set of objects `K` in the given `Api` scope, you can use [`ListParams::default`].
+    #[must_use]
+    pub fn new(api: Api<K>) -> Self {
+        Self {
+            api,
+            listparams: None,
+            backoff: None,
+        }
+    }
+
+    // start the watcher and filter out backoff errors from the stream for a while
+    fn start(self) -> impl Stream<Item = Result<Event<K>>> {
+        let backoff = self.backoff.unwrap_or_else(|| backoff::ExponentialBackoff {
+            max_elapsed_time: Some(std::time::Duration::from_secs(60 * 10)),
```

Comment:
This feels like it's in the dead zone where we're likely to end up with deadlocks and other weird issues. Either this should be short enough to basically say "no, we just crash on failures", or it should be infinite.

Reply:
if we wanted to always crash on failures, we wouldn't need backoff here at all though. if things haven't recovered by 10 minutes, then it seems like a reasonable time to crash to me - it's unlikely to get better - and crashloopbackoff will cause a full retry cycle anyway. maybe it can be tweaked within one order of magnitude (e.g. anything between 1min -> 1 hour seems reasonable to me), but kind of want the default to crash at some point unless users ask for infinite retry.

Reply:
This is using client-go values now.